// @nextcloud/vue/dist/chunks/NcAppNavigationSettings-C8YaW_Kj.cjs
require('../assets/NcMentionBubble-UsTMm5VD.css');
require('../assets/NcAppNavigationSettings-_Qi3rUlw.css');
"use strict";
const mdi = require("./mdi-B_QN3oas.cjs");
const components = require("@vueuse/components");
const NcIconSvgWrapper = require("./NcIconSvgWrapper-CrnN5Sq5.cjs");
const Mixins_clickOutsideOptions = require("../Mixins/clickOutsideOptions.cjs");
require("../Composables/useIsFullscreen.cjs");
require("../Composables/useIsMobile.cjs");
require("../Composables/useIsDarkTheme.cjs");
require("@nextcloud/router");
require("../Functions/isDarkTheme.cjs");
;/* empty css                                                                           */
const _pluginVue2_normalizer = require("./_plugin-vue2_normalizer-V0q-tHlQ.cjs");
require("escape-html");
require("striptags");
require("vue");
require("@nextcloud/auth");
require("@nextcloud/axios");
require("@nextcloud/capabilities");
const legacy = require("./legacy-KBXhWdRy.cjs");
const _l10n = require("./_l10n-RKohWu5d.cjs");
_l10n.register(_l10n.t45);
const _sfc_main = {
  directives: {
    ClickOutside: components.vOnClickOutside
  },
  components: {
    NcIconSvgWrapper: NcIconSvgWrapper.NcIconSvgWrapper
  },
  mixins: [
    Mixins_clickOutsideOptions
  ],
  props: {
    name: {
      type: String,
      required: false,
      default: _l10n.t("Settings")
    }
  },
  setup() {
    return {
      isLegacy32: legacy.isLegacy32,
      mdiCog: mdi.mdiCog,
      mdiCogOutline: mdi.mdiCogOutline
    };
  },
  data() {
    return {
      open: false
    };
  },
  computed: {
    clickOutsideConfig() {
      return [
        this.closeMenu,
        this.clickOutsideOptions
      ];
    }
  },
  methods: {
    toggleMenu() {
      this.open = !this.open;
    },
    closeMenu() {
      this.open = false;
    }
  }
};
var _sfc_render = function render() {
  var _vm = this, _c = _vm._self._c;
  return _c("div", { directives: [{ name: "click-outside", rawName: "v-click-outside", value: _vm.clickOutsideConfig, expression: "clickOutsideConfig" }], class: { open: _vm.open }, attrs: { "id": "app-settings" } }, [_c("div", { attrs: { "id": "app-settings__header" } }, [_c("button", { staticClass: "settings-button", attrs: { "type": "button", "aria-expanded": _vm.open ? "true" : "false", "aria-controls": "app-settings__content" }, on: { "click": _vm.toggleMenu } }, [_c("NcIconSvgWrapper", { staticClass: "settings-button__icon", attrs: { "path": _vm.isLegacy32 ? _vm.mdiCog : _vm.mdiCogOutline } }), _c("span", { staticClass: "settings-button__label" }, [_vm._v(_vm._s(_vm.name))])], 1)]), _c("Transition", { attrs: { "name": "slide-up" } }, [_c("div", { directives: [{ name: "show", rawName: "v-show", value: _vm.open, expression: "open" }], attrs: { "id": "app-settings__content" } }, [_vm._t("default")], 2)])], 1);
};
var _sfc_staticRenderFns = [];
var __component__ = /* @__PURE__ */ _pluginVue2_normalizer.normalizeComponent(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  "3862a820"
);
const NcAppNavigationSettings = __component__.exports;
exports.NcAppNavigationSettings = NcAppNavigationSettings;
